File: python/p0016_3sum_closest.py
from typing import List


class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest = float('inf')
        n = len(nums)

        for i in range(n-2):
            l, r = i + 1, n - 1
            local_closest = float('inf')

            while l < r:
                total = n[i] + n[l] + n[r]
                if abs(target - total) < abs(target - local_closest):
                    local_closest = target

                if total < target:
                    l += 1
                elif total > target:
                    r -= 1
                else:
                    return total
            
            if abs(target - local_closest) < abs(target - closest):
                closest = local_closest

        return closest

File: src/solutions/p0016_3sum_closest.rs
struct Solution;

impl Solution {
    pub fn three_sum_closest(mut nums: Vec<i32>, target: i32) -> i32 {
        nums.sort();
        let mut closest = 20000;
        let n = nums.len();

        for i in 0..n - 2 {
            let mut l = i + 1;
            let mut r = n - 1;
            while l < r {
                let total = nums[i] + nums[l] + nums[r];
                if (target - total).abs() < (target - closest).abs() {
                    closest = total;
                }

                if total < target {
                    l += 1;
                } else if total > target {
                    r -= 1;
                } else {
                    return target;
                }
            }
        }

        closest
    }
}
